core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java

```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
 
   static Metrics fromWriter(DatumWriter<?> datumWriter, Schema schema, long numRecords,
                             MetricsConfig inputMetricsConfig) {
-    // TODO will populate in following PRs if datum writer is a MetricsAwareDatumWriter
-    return new Metrics(numRecords, null, null, null);
+    if (!(datumWriter instanceof MetricsAwareDatumWriter)) {
+      return new Metrics(numRecords, null, null, null, null);
+    }
+
+    MetricsAwareDatumWriter<?> metricsAwareDatumWriter = (MetricsAwareDatumWriter<?>) datumWriter;
+    MetricsConfig metricsConfig;
+    if (inputMetricsConfig == null) {
+      metricsConfig = MetricsConfig.getDefault();
+    } else {
+      metricsConfig = inputMetricsConfig;
+    }
+
+    Map<Integer, Long> valueCounts = new HashMap<>();
+    Map<Integer, Long> nullValueCounts = new HashMap<>();
+    Map<Integer, Long> nanValueCounts = new HashMap<>();
+    Map<Integer, ByteBuffer> lowerBounds = new HashMap<>();
+    Map<Integer, ByteBuffer> upperBounds = new HashMap<>();
+
+    metricsAwareDatumWriter.metrics().forEach(metrics -> {
+      String columnName = schema.findColumnName(metrics.id());
+      MetricsModes.MetricsMode metricsMode = metricsConfig.columnMode(columnName);
+      if (metricsMode == MetricsModes.None.get()) {
+        return;
+      }
+
+      valueCounts.put(metrics.id(), metrics.valueCount());
+      nullValueCounts.put(metrics.id(), metrics.nullValueCount());
+      Type type = schema.findType(metrics.id());
+
+      if (type.typeId() == Type.TypeID.FLOAT || type.typeId() == Type.TypeID.DOUBLE) {
+        nanValueCounts.put(metrics.id(), metrics.nanValueCount());
+      }
+
+      if (metricsMode == MetricsModes.Counts.get()) {
+        return;
+      }
+
+      updateLowerBound(metrics, type, metricsMode).ifPresent(lowerBound -> lowerBounds.put(metrics.id(), lowerBound));
+      updateUpperBound(metrics, type, metricsMode).ifPresent(upperBound -> upperBounds.put(metrics.id(), upperBound));
+    });
+
+    return new Metrics(numRecords, null,
+        valueCounts, nullValueCounts, nanValueCounts, lowerBounds, upperBounds);
+  }
+
+  private static Optional<ByteBuffer> updateLowerBound(FieldMetrics metrics, Type type,
+                                                       MetricsModes.MetricsMode metricsMode) {
+    if (metrics.lowerBound() == null) {
+      return Optional.empty();
+    }
+
+    Object lowerBound = metrics.lowerBound();
+    if (metricsMode instanceof MetricsModes.Truncate) {
+      MetricsModes.Truncate truncateMode = (MetricsModes.Truncate) metricsMode;
+      int truncateLength = truncateMode.length();
+      switch (type.typeId()) {
+        case STRING:
+          lowerBound = UnicodeUtil.truncateStringMin(
+              Literal.of((CharSequence) metrics.lowerBound()), truncateLength).value();
```

Comment:
Why not use `UnicodeUtil.truncateStringMin(CharSequence, int)` instead of the `Literal` method?

Reply:
Ah, I see that the `min` and `max` methods take literals. I think it would be better to refactor those methods to expose non-Literal implementations instead of creating the literals and unwrapping the result.